### PR TITLE
Updating supported Docker commands, adding Docker Compose file options.

### LINKED ISSUE #839: Document Docker events
Per https://github.com/vmware/vic/issues/81. This probably requires info in both of the App Dev and vSphere Admin docs.

### LINKED ISSUE #1696: Document docker ps filtering options
https://github.com/vmware/vic/issues/1676 adds filtering options to docker ps. This potentially needs to be called out in the doc.

### LINKED ISSUE #3762: Document the docker compose commands that currently don't work
Per https://github.com/vmware/vic/issues/3635. This definitely deserves a big table in the developer doc.

### LINKED ISSUE #3848: Document support for image filtering
Per https://github.com/vmware/vic/issues/2248, which is implemented in PR https://github.com/vmware/vic/pull/3847.

### LINKED ISSUE #3849: Document support for volume ls filtering
Per https://github.com/vmware/vic/issues/1718, which is implemented in PR https://github.com/vmware/vic/pull/3834.

### LINKED ISSUE #3855: Document docker images tag
Per https://github.com/vmware/vic/issues/372 (parent issue for `images` options, and https://github.com/vmware/vic/issues/1120 specifically for `tag`.

### LINKED ISSUE #4007: Update the user docs to specify the next version of VIC only supports compose format 3
As a user of VIC, I should be able to find out which version of the compose file format and compose binary version that VIC supports by reading the user docs.

Docker-compose 1.10 and up
Compose file format 3 and up

Best effort to use docker-compose binary 1.9.0

### PR DESCRIPTION
**Pull request by stuclem**
Also did some general clean up in the Supported Docker Commands tables.

- Fixes #3762 
- Fixes #839 
- Fixes #1696 
- Fixes #3849 
- Fixes #3848 
- Fixes #3855 
- Fixes #4007

### CODE REVIEW

**Comment by stuclem** [comment]
@sflxn and @chengwang86 can you please review these changes for 0.9? Thanks!

**Comment by stuclem** [comment]
Here are the updated topics in their entirety:
- https://github.com/stuclem/vic/blob/docker/doc/user_doc/vic_app_dev/container_operations.md
- https://github.com/stuclem/vic/blob/docker/doc/user_doc/vic_app_dev/docker_compose_file_options.md

File: doc/user_doc/vic_app_dev/container_operations.md [hunk 1]
```diff
@@ -29,12 +29,12 @@
...
 |`commit`|[Create a new image from a container’s changes](https://docs.docker.com/engine/reference/commandline/commit/)|No|
 |`history`|[Show the history of an image](https://docs.docker.com/engine/reference/commandline/history/)|No|
-|`images`|[Images](https://docs.docker.com/engine/reference/commandline/images/)|Yes. vSphere Integrated Container Engine currently supports `--no-trunc` and `-q/--quiet`|
+|`images`|[Images](https://docs.docker.com/engine/reference/commandline/images/)|Yes. Supports `--filter`, `--no-trunc`, and `--quiet`|
 |`import`|[Import the contents from a tarball to create a filesystem image](https://docs.docker.com/engine/reference/commandline/import/)|No|
 |`load`|[Load an image from a tar archive or STDIN](https://docs.docker.com/engine/reference/commandline/load/)|No|
-|`rmi`|[Remove a Docker image](https://docs.docker.com/engine/reference/commandline/rmi/)|Yes. vSphere Integrated Container Engine does not yet support tag or untag operations, or any options.|
+|`rmi`|[Remove a Docker image](https://docs.docker.com/engine/reference/commandline/rmi/)|Yes. Does not yet support `tag` or `untag` operations, or any options.|
```
Comment (chengwang86) [inline doc/user_doc/vic_app_dev/container_operations.md]: Does `rmi` have `tag` or `untag` operations?

File: doc/user_doc/vic_app_dev/container_operations.md [hunk 2]
```diff
@@ -45,23 +45,23 @@
...
+|`rm`|[Remove a container](https://docs.docker.com/engine/reference/commandline/rm/)|Yes. Removes associated anonymous and regular volumes. Supports the `--force` option and the `name` parameter.  `docker rm -v` and `docker rm -f` are not supported. To view volumes attached to a container that is removed, use `docker volume ls` and `docker volume inspect <id>`. If you continually invoke `docker create` to make more anonymous volumes, those volumes are left behind after each subsequent removal of that container.|
```
Comment (chengwang86) [inline doc/user_doc/vic_app_dev/container_operations.md]: I think we do support `docker rm -f`

File: doc/user_doc/vic_app_dev/container_operations.md [hunk 3]
```diff
@@ -100,43 +100,45 @@ For more information about volume operations with vSphere Integrated Containers
 | **Command** | **Docker Reference** | **Supported** |
 | --- | --- | --- |
 |`volume create`|[Create a volume](https://docs.docker.com/engine/reference/commandline/volume_create/)|Yes|
-|`volume inspect`|[Information about a volume](https://docs.docker.com/engine/reference/commandline/volume_inspect/)|Yes, use with docker compose.|
-|`volume ls`|[List volumes](https://docs.docker.com/engine/reference/commandline/volume_ls/)|Yes. `--filter` is not supported.|
+|`volume inspect`|[Information about a volume](https://docs.docker.com/engine/reference/commandline/volume_inspect/)|Yes. Use with docker compose.|
```
Comment (chengwang86) [inline doc/user_doc/vic_app_dev/container_operations.md]: No need to mention "Use with docker compose".

File: doc/user_doc/vic_app_dev/container_operations.md [hunk 4]
```diff
@@ -100,43 +100,45 @@ For more information about volume operations with vSphere Integrated Containers
...
+| `bundle`  | [Generate a Distributed Application Bundle (DAB) from the Compose file](https://docs.docker.com/compose/reference/bundle/)| No |
 | `config`  | [Validate and view the compose file](https://docs.docker.com/compose/reference/config/)  | Yes  |
 | `create`  | [Create services](https://docs.docker.com/compose/reference/create/)  | Yes  |
 | `down`  | [Stop and remove containers, networks, images, and volumes](https://docs.docker.com/compose/reference/down/)  | Yes  |
-| `events`  |[Receive real time events from containers](https://docs.docker.com/compose/reference/events/)  | No. Depends on `docker events`.|
+| `events`  |[Receive real time events from containers](https://docs.docker.com/compose/reference/events/)  | Yes. Supports passive Docker events for containers and images. When you perform `docker compose stop`, regular Docker Compose reports the `kill > die > stop` events, however vSphere Integrated Containers Engine reports `die > stop`. Regular Docker Compose reports the name of the image for any container operation, however vSphere Integrated Containers Engine reports `image=sha256:<hashtag>`. Does not yet support events for volumes or networks.|
```
Comment (chengwang86) [inline doc/user_doc/vic_app_dev/container_operations.md]: Can we just say this here:

Yes. Supports passive Docker events for containers and images. Does not yet support events for volumes or networks.

File: doc/user_doc/vic_app_dev/docker_compose_file_options.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+|`security-opt` |[Override the default labeling scheme for containers](https://docs.docker.com/compose/compose-file/compose-file-v2/#securityopt) | No. This option only applies to Windows containers, which vSphere Integrated Containers Engine does not support. |
+|`stop-signal` |[Sets an alternative signal to stop the container.](https://docs.docker.com/compose/compose-file/compose-file-v2/#stopsignal) | Yes |
+|`stop-grace-period` |[Specify how long to wait stopping a container](https://docs.docker.com/compose/compose-file/compose-file-v2/#stopgraceperiod) | No |
+|`sysctls` |[Kernel parameters to set in the container](https://docs.docker.com/compose/compose-file/compose-file-v2/#sysctls) | No |
+|`ulimits` |[Override the default ulimits for a container](https://docs.docker.com/compose/compose-file/compose-file-v2/#ulimits) | No |
+|`userns_mode` |[Disables the user namespace](https://docs.docker.com/compose/compose-file/compose-file-v2/#usernsmode) | No |
+|`volumes`,`volume_driver` |[xxx](https://docs.docker.com/compose/compose-file/compose-file-v2/#volumes-volumedriver) | Yes |
```
Comment (chengwang86) [inline doc/user_doc/vic_app_dev/docker_compose_file_options.md]: xxx

File: doc/user_doc/vic_app_dev/docker_compose_file_options.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+|`volumes_from` |[Mount volumes from another service or container](https://docs.docker.com/compose/compose-file/compose-file-v2/#volumesfrom)| No |
+
+
+ The following [Docker run options](https://docs.docker.com/compose/compose-file/compose-file-v2/#cpushares-cpuquota-cpuset-domainname-hostname-ipc-macaddress-memlimit-memswaplimit-memswappiness-oomscoreadj-privileged-readonly-restart-shmsize-stdinopen-tty-user-workingdir) are supported if their [`docker run`](container_operations.md#docker_run) counterpart is supported: `security_opt`, `stop_grace_period`, `stop_signal`, `sysctls`, `ulimits`, `userns_mode`, `cpu_shares`, `cpu_quota`, `cpuset`, `domainname`, `hostname`, `ipc`, `mac_address`, `mem_limit`, `memswap_limit`, `oom_score_adj`, `privileged`, `read_only`, `restart`, `shm_size`, `stdin_open`, `tty`, `user`, `working_dir`.
+
+<a id="volume"></a>
+## Volume Configuration Options ##
+
+NOTE: vSphere Integrated Containers 0.9 does support shared volumes. You can use these options for containers that do not share volumes.
```
Comment (chengwang86) [inline doc/user_doc/vic_app_dev/docker_compose_file_options.md]: vSphere Integrated Containers 0.9 does **not** support shared volumes

File: doc/user_doc/vic_app_dev/docker_compose_file_options.md [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+<a id="network"></a>
+## Network Configuration Options ##
+
+| **Option** | **Compose File Reference** | **Supported** |
+| --- | --- | --- |
+|`driver` | [Specify driver to use for this network](https://docs.docker.com/compose/compose-file/compose-file-v2/#driver-1) | Yes |
+|`driver_opts` | [Specify options to pass to the driver for this network](https://docs.docker.com/compose/compose-file/compose-file-v2/#driveropts-1) | No |
+|`enable_ipv6` | [Enables IPv6](https://docs.docker.com/compose/compose-file/compose-file-v2/#enableipv6) | No. vSphere Integrated Containers Engine does not support IPv6. |
+|`ipam`| [Specify custom IPAM configuration](https://docs.docker.com/compose/compose-file/compose-file-v2/#ipam) | No. Depends on `docker network create --ipam`. |
```
Comment (chengwang86) [inline doc/user_doc/vic_app_dev/docker_compose_file_options.md]: Sorry it's my fault. We do support `ipam` in compose for 0.9.

File: doc/user_doc/vic_app_dev/container_operations.md (hunk 1 above)
Comment (stuclem) [inline doc/user_doc/vic_app_dev/container_operations.md]: It seems not: https://docs.docker.com/engine/reference/commandline/rmi/

I can only imagine that this is a copy/paste error. Removed.

File: doc/user_doc/vic_app_dev/container_operations.md (hunk 2 above)
Comment (stuclem) [inline doc/user_doc/vic_app_dev/container_operations.md]: Removed from the list.

File: doc/user_doc/vic_app_dev/container_operations.md (hunk 3 above)
Comment (stuclem) [inline doc/user_doc/vic_app_dev/container_operations.md]: Removed.

File: doc/user_doc/vic_app_dev/container_operations.md (hunk 4 above)
Comment (stuclem) [inline doc/user_doc/vic_app_dev/container_operations.md]: Done. I thought that I had fixed that already, but it seems not.

File: doc/user_doc/vic_app_dev/docker_compose_file_options.md (hunk 6 above)
Comment (stuclem) [inline doc/user_doc/vic_app_dev/docker_compose_file_options.md]: Oops! Fixed.

File: doc/user_doc/vic_app_dev/docker_compose_file_options.md (hunk 5 above)
Comment (stuclem) [inline doc/user_doc/vic_app_dev/docker_compose_file_options.md]: Fixed

File: doc/user_doc/vic_app_dev/docker_compose_file_options.md (hunk 7 above)
Comment (stuclem) [inline doc/user_doc/vic_app_dev/docker_compose_file_options.md]: Fixed. Thanks @chengwang86 !

**Comment by chengwang86** [comment]
lgtm

**Pull request merged**